### PR TITLE
Add typed spans working document

### PR DESCRIPTION
**Pull request by danielkhan**

### CODE REVIEW

**Comment by bogdandrutu** [comment]
@danielkhan thanks for the PR, you need to sign the CLA.

File: work_in_progress/typedspans/README.md [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+there is no equivalent in OC. Additionally, some attributes that may enrich the
+span further are proposed.
+
+**Bold** = mandatory.
+
+#### HttpClient
+
+<table>
+  <tr>
+   <td><strong>CanonicalType = HTTP</strong>
```
Comment (flands) [inline work_in_progress/typedspans/README.md]: Would this be an attribute or first-class citizen on span?

File: work_in_progress/typedspans/README.md [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+   <td>http.route
+   </td>
+  </tr>
+  <tr>
+   <td>http.user_agent
+   </td>
+  </tr>
+  <tr>
+   <td>Proposed additions:
+parameters, requestHeaders, responseHeaders
```
Comment (flands) [inline work_in_progress/typedspans/README.md]: I assume these would be attributes? If so then should standardize on format -- OpenTracing one is pretty good: all lowercase, underscore between words, period between types.

File: work_in_progress/typedspans/README.md [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+  <tr>
+   <td><strong>CanonicalType = DB</strong>
+   </td>
+  </tr>
+  <tr>
+   <td><strong>SpanKind = CLIENT</strong>
+   </td>
+  </tr>
+  <tr>
+   <td><strong>db.instance</strong>
```
Comment (flands) [inline work_in_progress/typedspans/README.md]: These seems contrary to OpenTracing. The mandatory field should be `db.type`. Why `db.instance`?

File: work_in_progress/typedspans/README.md [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+   </td>
+  </tr>
+  <tr>
+   <td><strong>component</strong>
```
Comment (flands) [inline work_in_progress/typedspans/README.md]: Why is this required?

File: work_in_progress/typedspans/README.md [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+   </td>
+  </tr>
+  <tr>
+   <td><strong>peer.*</strong>
```
Comment (flands) [inline work_in_progress/typedspans/README.md]: If required, all options should be explicitly defined. Why are any of these required?

File: work_in_progress/typedspans/README.md [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+The following canonical types should be implemented.
+The attributes are taken from the OC spec first with a fallback to the OT spec if
+there is no equivalent in OC. Additionally, some attributes that may enrich the
+span further are proposed.
+
+**Bold** = mandatory.
+
+#### HttpClient
+
+<table>
```
Comment (yurishkuro) [inline work_in_progress/typedspans/README.md]: Can we do this in plain markdown? The tables have a single column, can be easily represented with bullet lists.

File: work_in_progress/typedspans/README.md (hunk 2 above)
Comment (yurishkuro) [reply work_in_progress/typedspans/README.md]: What is the difference between "proposed" and "not required"?

File: work_in_progress/typedspans/README.md (hunk 1 above)
Comment (danielkhan) [reply work_in_progress/typedspans/README.md]: > Would this be an attribute or first-class citizen on span?

If you are referring to `CanonicalType`, first class citizen would be the preference.

File: work_in_progress/typedspans/README.md (hunk 2 above)
Comment (danielkhan) [reply work_in_progress/typedspans/README.md]: > I assume these would be attributes? If so then should standardize on format -- OpenTracing one is pretty good: all lowercase, underscore between words, period between types.

This proposal was created by combining OpenCensus and OpenTracing naming and I think that there will be a general discussion on attribute naming rules for Open Telemetry. So this is more about the general rationale behind typed spans and which attributes would make sense.

Comment (danielkhan) [reply work_in_progress/typedspans/README.md]: > What is the difference between "proposed" and "not required"?

This is meant to be an addition to what exists already in OpenCensus and OpenTracing.

File: work_in_progress/typedspans/README.md (hunk 3 above)
Comment (danielkhan) [reply work_in_progress/typedspans/README.md]: Maybe this was a mistake in reading the spec. Looking at https://github.com/opentracing/specification/blob/master/semantic_conventions.md#database-client-calls it seems as if both should be required. Would you agree?

File: work_in_progress/typedspans/README.md (hunk 4 above)
Comment (danielkhan) [reply work_in_progress/typedspans/README.md]: Because it is in Dynatrace's data model. We would map it to our dbVendor field.

File: work_in_progress/typedspans/README.md (hunk 5 above)
Comment (danielkhan) [reply work_in_progress/typedspans/README.md]: They are required to map them to a concrete instance in a model that might be external to OpenTelemetry.

File: work_in_progress/typedspans/README.md (hunk 2 above)
Comment (discostu105) [reply work_in_progress/typedspans/README.md]: Maybe we should not have these in the initial proposal. They haven't been part of OT/OC (in this form), and they have an additional complexity (they are key/value pair lists). Maybe we should discuss them separately.

**Comment by danielkhan** [comment]
> @danielkhan thanks for the PR, you need to sign the CLA.

Yes we are waiting for our legal team to take care of that.

**Comment by tylerbenson** [comment]
Can anyone offer insight as to how we expect this would be implemented? For example, would there be many different sub-classes of Span, each with different sets of fields?  Or perhaps a "typed attribute" field inside Span that these would be set on?

**Comment by discostu105** [comment]
@tylerbenson IMHO this work should be split in two parts. 

1: Just documentation of what attributes exist and which are supposed to be on certain kinds (more like OT "semantic conventions", but with the canonical type notion). Not technically binding, but rather recommendations. Why? It has value both on implementing instrumentation as well as on interpreting data (analytics backend).

2: Typed spans API. This is a different discussion (we should do separately :)). IMHO we should keep generic span classes, with generic span attributes. But we could add typed "Builder" classes that would help a user to set the correct attributes (via dedicated API's for each attribute).

File: work_in_progress/typedspans/README.md (hunk 2 above)
Comment (danielkhan) [reply work_in_progress/typedspans/README.md]: I changed the structure but left them in there so that this information does not get lost. There will be subsequent discussions on which fields should be in the spec eventually.

File: work_in_progress/typedspans/README.md (hunk 4 above)
Comment (danielkhan) [reply work_in_progress/typedspans/README.md]: The latest version of the doc does not distinguish between required and optional as this should be a dedicated discussion by type.

File: work_in_progress/typedspans/README.md (hunk 6 above)
Comment (danielkhan) [reply work_in_progress/typedspans/README.md]: Done

**Comment by SergeyKanzhelev** [comment]
@danielkhan I think it's a good idea to merge this document in and then start transferring pieces of it to the folder outside the `work_in_progress` folder. We started by copying open census and open tracing documents into subfolders of this folder and when specific item is described in the document like https://github.com/open-telemetry/opentelemetry-specification/blob/master/work_in_progress/gRPC/gRPC.md or https://github.com/open-telemetry/opentelemetry-specification/blob/master/work_in_progress/database/database.md - those attributes are removed from the originally copied specs.

Does this approach make sense for you?

**Comment by danielkhan** [comment]
> @danielkhan I think it's a good idea to merge this document in and then start transferring pieces of it to the folder outside the `work_in_progress` folder. 
> [...]
> Does this approach make sense for you?

Absolutely. That sounds like a good approach to get the ball rolling.

File: work_in_progress/typedspans/README.md (hunk 1 above)
Comment (SergeyKanzhelev) [reply work_in_progress/typedspans/README.md]: created an issue for this: https://github.com/open-telemetry/opentelemetry-specification/issues/16

**Comment by SergeyKanzhelev** [comment]
I'm going to merge it as this is work in progress. Basically an accumulation of many documents. Documents that will require a bigger consensus will be placed outside of this folder.

**Comment by SergeyKanzhelev** [comment]
> @tylerbenson IMHO this work should be split into two parts.

@discostu105 perhaps there should be the separate issues for those separate discussions than
